contracts/ProductNft/ProductNft.sol

```diff
@@ -0,0 +1,170 @@
+        tokenAmounts = new uint[](amount);
+        for (uint256 i = 0; i < amount;) {
+            tokenAmounts[i] = i + 1;
+            unchecked {
+                i++;
+            }
+        }
+    }
+
+    /**
+    *@dev   A counter is used to track the token Ids that have been minted.
+    *       A token Id will be returned and turned into an array. 
+    *@param rarity The rarity defines at what point the count starts.
+    *       Using requirements, rarities have been assigned a set of tokenIds. 
+    */
+    function _countBasedOnRarity(uint8 rarity) internal returns (uint256 tokenId) {
+        if(rarity == rarityRarest) {
+            rarestTokenIds.increment();
+            tokenId = rarestTokenIds.current();
+            require(tokenId <= startRarerTokenIdIndex, "Mycelia NFTs are sold out");
+            return tokenId;   
+        } else if (rarity == rarityRarer) {
+            rarerTokenIds.increment();
+            tokenId = startRarerTokenIdIndex + rarerTokenIds.current();
+            require(tokenId >= startRarerTokenIdIndex && tokenId <= startRareTokenIdIndex, "Diamond NFTS are sold out");
+            return tokenId;
+        } else if (rarity == rarityRare) {
+            rareTokenIds.increment();
+            tokenId = startRareTokenIdIndex + rareTokenIds.current();
+            require(tokenId <= totalAmountOfTokenIds, "Silver NFTs are sold out");
+            return tokenId;
+        }
+    }
+
+    function _setURI(uint256 tokenId) internal {
+            _URIS[tokenId] = _URI(tokenId);
+    }
+
+    /**
+    *@dev   For ERC1155's batch mint function, a an array of token Ids
+    *       with equal length to the array of amounts is needed.  
+    *@param rarity The rarity defines at what point the count starts.
+    *@param amount the given amount that will be turned into an array as
+    *       the _turnAmountIntoArray function is used. 
+    *       For each entry in the amounts array, the counter will be used 
+    *       to generate the next token Id and to store that Id in an array.
+    */
+    function _turnTokenIdsIntoArray(uint8 rarity, uint16 amount) internal returns (uint256[] memory tokenIdArray) {
+        tokenIdArray = new uint[](_turnAmountIntoArray(amount).length); 
+        for (uint16 i = 0; i < _turnAmountIntoArray(amount).length;) { 
+            uint256 currentTokenId = _countBasedOnRarity(rarity);
+            tokenIdArray[i] = currentTokenId;
+            _setURI(currentTokenId);
+            unchecked {
+                i++;
+            }  
+        }
+    }
+
+    /**
+    *@dev   This minting function allows the minting of Rarest tokenIds.
+    *@param amount: Every call will recursively increment the tokenId 
+    *       depending on the amount of tokens the user wants to batch mint.
+    *       These tokenIds are associated with the Mycelia rarity. 
+    *       This function can be called for 1.5 ETH.
+    */
+    function rarestBatchMint(uint16 amount) public payable {
```

Comment:
Can we add some events to these 3 batchMint functions, and a corresponding test?